src/System.CommandLine/Invocation/InvocationPipeline.cs

```diff
@@ -24,11 +24,32 @@ internal static async Task<int> InvokeAsync(ParseResult parseResult, Cancellatio
                 {
                     for (int i = 0; i < parseResult.NonexclusiveActions.Count; i++)
                     {
-                        await parseResult.NonexclusiveActions[i].InvokeAsync(parseResult, cts.Token);
+                        var action = parseResult.NonexclusiveActions[i];
+
+                        switch (action)
+                        {
+                            case SynchronousCliAction syncAction:
+                                syncAction.Invoke(parseResult);
```

Comment:
> If cancellation is triggered by console ^C, then it would be useful to inform sync actions about it, too.

How? Any why anyone should use sync overload to get cancellation support which is one of the main reasons of using async?

Reply:
I was looking for a place to put the same comment @KalleOlaviNiemitalo 's puts here. It is reasonable to have sync that supports cancellation via events. If we do not support that, we should make a clear decision on preferring the benefit of a single cancellation strategy to the benefit of not having to mess with tasks when they are otherwise unnecessary.

Reply:
It's also reasonable to use `CancellationToken` in synchronous APIs. The idea was suggested by API review.

Reply:
> CPU-bound work but with temporary files and directories that should be deleted when canceled.

This can be achieved with `FileOptions.DeleteOnClose` which deletes the file when the handle is released.

> It's also reasonable to use `CancellationToken` in synchronous APIs. 

It's definitely possible, but I am not sure whether the users would be able to take advantage of that.

With async code, everyone just needs to propagate the token to the async methods they are using and the compiler emits a warning when they don't.

At some point of time, the token is passed to one of the `System*` methods that implement cancellation: they monitor the change of the token state and when it's requested, they perform actual cancellation by for example asking the OS to cancell given IO operaiton.

With sync code, the users would be left on their own. They would need to constantly monitor the status of the token, and when cancellation is requested, handle it on their own and throw an exception.
But they could interrupt only some of the operations, namely only their own CPU-bound computations. They could not cancel IO operations, as they would be using the sync ones. Or at least they should, as sync should not be mixed with async. If they end up calling async methods from sync methods just to pass the cancellation token and perform blocking wait on the task, the design leads them into failure.

So it's possible, but I don't believe that it's the right direction as it would increase complexity and almost never provide the required value.

Reply:
>> CPU-bound work but with temporary files and directories that should be deleted when canceled.

> This can be achieved with `FileOptions.DeleteOnClose` which deletes the file when the handle is released.

This is just one example. Not all resources (e.g. processes, cloud resources) will have APIs for automatic cleanup.